src/inmanta/data/__init__.py

```diff
@@ -213,6 +233,296 @@ def ensure_boundary_type(
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "ResourceHistoryOrder":
+        database_order = super().parse_from_string(sort)
+        return ResourceHistoryOrder(database_order.order_by_column, database_order.order)
+
+
+class ResourceLogOrder(DatabaseOrder):
+    """ Represents the ordering by which resource logs should be sorted"""
+
+    valid_sort_columns = {"timestamp": datetime.datetime}
+
+    @classmethod
+    def validator_dataclass(cls) -> Type["BaseDocument"]:
+        return ResourceAction
+
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "ResourceLogOrder":
+        database_order = super().parse_from_string(sort)
+        return ResourceLogOrder(database_order.order_by_column, database_order.order)
+
+
+class CompileReportOrder(DatabaseOrder):
+    """ Represents the ordering by which compile reports should be sorted"""
+
+    valid_sort_columns = {"requested": datetime.datetime}
+
+    @classmethod
+    def validator_dataclass(cls) -> Type["BaseDocument"]:
+        return Compile
+
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "CompileReportOrder":
+        database_order = super().parse_from_string(sort)
+        return CompileReportOrder(database_order.order_by_column, database_order.order)
+
+
+class BaseQueryBuilder(ABC):
+    """Provides a way to build up a sql query from its parts"""
+
+    def __init__(
+        self,
+        select_clause: Optional[str] = None,
+        from_clause: Optional[str] = None,
+        filter_statements: Optional[List[str]] = None,
+        values: Optional[List[object]] = None,
+    ) -> None:
+        self.select_clause = select_clause
+        self._from_clause = from_clause
+        self.filter_statements = filter_statements or []
+        self.values = values or []
+
+    def _join_filter_statements(self, filter_statements: List[str]) -> str:
+        """ Join multiple filter statements """
+        if filter_statements:
+            return "WHERE " + " AND ".join(filter_statements)
+        return ""
+
+    def from_clause(self, from_clause: str):
+        """ Set the from clause of the query"""
+        self._from_clause = from_clause
+        return self
+
+    @property
+    def offset(self):
+        """ The current offset of the values to be used for filter statements"""
+        return len(self.values) + 1
+
+    def filter(self, filter_statements: List[str], values: List[object]) -> "BaseQueryBuilder":
+        """ Add filters to the query """
+        self.filter_statements += filter_statements
+        self.values += values
+        return self
+
+    @abstractmethod
+    def build(self) -> Tuple[str, List[object]]:
+        """ Builds up the full query string, and the parametrized value list, ready to be executed """
+        pass
+
+
+class SimpleQueryBuilder(BaseQueryBuilder):
+    """ A query builder suitable for most queries """
+
+    def __init__(
```

Comment:
All arguments need documentation

Reply:
Added it in the docstring of the `__init__` method